paddle/phi/kernels/gpu/layer_norm_kernel.cu

```diff
@@ -13,14 +13,372 @@
 
 #include "paddle/phi/kernels/layer_norm_kernel.h"
-
+#include "gflags/gflags.h"
 #include "paddle/fluid/operators/layer_norm_kernel.cu.h"
 #include "paddle/phi/backends/gpu/gpu_context.h"
 #include "paddle/phi/core/kernel_registry.h"
 #include "paddle/phi/kernels/funcs/layer_norm_util.h"
 
+DECLARE_bool(use_fast_math);
+namespace ops = paddle::operators;
+
 namespace phi {
 
+#ifdef PADDLE_WITH_CUDA
+template <typename U>
+__device__ void WelfordOnline(U val, U *mean, U *square, U *count) {
+  *count += 1;
+  U delta1 = val - *mean;
+  *mean += delta1 / (*count);
+  U delta2 = val - *mean;
+  *square += delta1 * delta2;
+}
+
+template <typename U>
+__device__ void WelfordOnline(
+    U b_mean, U b_square, U b_cnt, U *mean, U *square, U *count) {
+  if (b_cnt == 0) {
+    return;
+  }
+
+  U new_cnt = *count + b_cnt;
+  U nb_n = b_cnt / new_cnt;
+  U delta = b_mean - *mean;
+  *mean += delta * nb_n;
+  *square += b_square + delta * delta * (*count) * nb_n;
+  *count = new_cnt;
+}
+
+template <typename U>
+__device__ void WelfordWarpAllReduce(U *mean, U *square, U *count) {
+  constexpr int kWarpSize = 32;
+#pragma unroll
+  for (int mask = 1; mask < kWarpSize; mask *= 2) {
+    U b_mean = __shfl_down_sync(0xffffffff, *mean, mask);
+    U b_square = __shfl_down_sync(0xffffffff, *square, mask);
+    U b_cnt = __shfl_down_sync(0xffffffff, *count, mask);
+    WelfordOnline<U>(b_mean, b_square, b_cnt, mean, square, count);
+  }
+
+  *mean = __shfl_sync(0xffffffff, *mean, 0, kWarpSize);
+  *square = __shfl_sync(0xffffffff, *square, 0, kWarpSize);
+  *count = __shfl_sync(0xffffffff, *count, 0, kWarpSize);
+}
+
+template <int VecSize>
+struct ThreadAssigner {
+  __device__ __forceinline__ int operator()(const int32_t cols,
+                                            const int32_t cols_per_thread) {
+    return cols_per_thread;
+  }
+};
+
+template <>
+struct ThreadAssigner<1> {
+  __device__ __forceinline__ int operator()(const int32_t cols,
+                                            const int32_t cols_per_thread) {
+    int cols_this_thread = cols_per_thread;
+    int last_y = (cols / cols_per_thread);
+    if (threadIdx.x == last_y) {
+      cols_this_thread = cols - cols_per_thread * last_y;
+    } else if (threadIdx.x > last_y) {
+      cols_this_thread = 0;
+    }
+    return cols_this_thread;
+  }
+};
+
+template <typename T, typename U, int VecSize>
+struct LayerNormDataReader {
+  __device__ __forceinline__ void operator()(const T *__restrict__ row_src,
+                                             U *buffer,
+                                             const int read_times,
+                                             const int cols_this_thread) {
+    using VecT = phi::AlignedVector<T, VecSize>;
+    const VecT *__restrict__ v_src =
+        reinterpret_cast<const VecT *__restrict__>(row_src);
+#pragma unroll
+    for (int i = 0; i < read_times; ++i) {
+      VecT temp_src = v_src[threadIdx.x * read_times + i];
```

Comment:
[ discussion ] If ```read_times``` were larger than 1, loading may not always be coalesced.

Reply:
yes, coalesced loading will not be accomplished, will be fixed later.